### DISCUSSION TITLE
Get the BackBuffer or a memory pointer for a bitmap

### DISCUSSION THREAD

**Discussion by Veriara** [SixLabors/ImageSharp#1375] [Q&A]
I'm converting an app that uses WPF to instead use ImageSharp. It renders the images on Cuda hardware via the ManagedCuda library. In order to pass the image data back to the host, I need to provide ManagedCuda with a place in the memory to assign to.

Does anyone know how I can achieve this?

**Comment by antonfirsov** [comment]
Is it sufficient to have a buffer for the time of an operation (inside a single, synchronous method)?

In that case, you can use [`image.TryGetSinglePixelSpan(out var span)`](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Image-1.html#SixLabors_ImageSharp_Image_1_TryGetSinglePixelSpan_System_Span__0___) then [pin the result](https://github.com/dotnet/corefxlab/blob/master/docs/specs/span.md#language-support), but in no case should you persist that pointer.

If you need a persistent reference to the buffer (to be stored in other objects), then it get's harder. You need to make sure to be familiar with the [.NET API-s for managing memory](https://docs.microsoft.com/en-us/archive/msdn-magazine/2018/january/csharp-all-about-span-exploring-a-new-net-mainstay#what-is-memoryt-and-why-do-you-need-it).

Then you can use `image.GetPixelMemoryGroup().Single()` to acquire a `Memory<TPixel>` to the buffer.

Note, that very large images are not contiguous in memory, but for images of size displayable on screen (below ~`4k*4k`) we plan to guarantee that the memory group will contain a single memory buffer, because of the GPU interop use-cases. See also: #1258
